### PR TITLE
Validate accountId and containerId in store server.

### PR DESCRIPTION
**Pull request by zzmao**
Validate accountId and containerId for GET and DELETE requests.
GET validation is done in AmbryRequests#handleGetRequests.
DELETE Validation is done in BlobStore#delete.
Authorization_Failure is added to StoreErrorCodes for both cases.

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java [hunk 1]
```diff
@@ -182,6 +182,14 @@
   @Default("115")
   public final int storeIndexPersistedEntryMinBytes;
 
+  @Config("store.get.authorization.check")
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java]: I don't think we need two different parameters for this - one should be enough

File: ambry-api/src/main/java/com.github.ambry/store/StoreKey.java [hunk 2]
```diff
@@ -37,6 +37,18 @@
    */
   public abstract String getID();
 
+  /**
+   * Get accountId of StoreKey
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKey.java]: Also add that it should be -1 if unknown

File: ambry-api/src/main/java/com.github.ambry/store/StoreKey.java [hunk 3]
```diff
@@ -37,6 +37,18 @@
...
+   * @return accountId
+   */
+  public abstract short getAccountId();
+
+  /**
+   * Get containerId of StoreKey
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKey.java]: Also add that it should be -1 if unknown

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 4]
```diff
@@ -22,6 +22,7 @@
 import com.github.ambry.clustermap.PartitionState;
 import com.github.ambry.clustermap.ReplicaId;
 import com.github.ambry.commons.ServerErrorCode;
+import com.github.ambry.commons.BlobId;
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: this import is no longer required.

File: ambry-server/src/main/java/com.github.ambry.server/ServerMetrics.java [hunk 5]
```diff
@@ -391,6 +393,10 @@ public ServerMetrics(MetricRegistry registry) {
     unExpectedStoreTTLError = registry.counter(MetricRegistry.name(AmbryRequests.class, "UnexpectedStoreTTLError"));
     unExpectedStoreFindEntriesError =
         registry.counter(MetricRegistry.name(AmbryRequests.class, "UnexpectedStoreFindEntriesError"));
+    getAuthorizationFailure =
+        registry.counter(MetricRegistry.name(AmbryRequests.class, "GetAuthorizationFailure"));
+    deleteAuthorizationFailure =
+        registry.counter(MetricRegistry.name(AmbryRequests.class, "DeleteAuthroizationFailure"));
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/ServerMetrics.java]: minor: Typo in "DeleteAuthroizationFailure" - "DeleteAuthorizationFailure

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 import com.github.ambry.config.StoreConfig;
 import com.github.ambry.utils.FileLock;
 import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: sorry for the back and forth but now that the validation check is all inside the store, I think we should move this function to this class as a private function.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java [hunk 7]
```diff
@@ -235,6 +236,15 @@ public StoreInfo get(List<? extends StoreKey> ids, EnumSet<StoreGetOptions> stor
         BlobReadOptions readInfo = index.getBlobReadInfo(key, storeGetOptions);
         readOptions.add(readInfo);
         indexMessages.put(key, readInfo.getMessageInfo());
+        // validate accountId and containerId
+        if (config.storeGetAuthorizationCheck && !Utils.validateAuthorization(
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: I think that we should always do the auth check but throw exception only if config is true, otherwise fire a metric. This way we can "test-run" this, make sure we haven't missed any corner cases and then turn on the config.

Something like
```
if (!authCheckOk()) {
  if (config) {
    // throw exception
  } else {
    // fire metric
  }
}
```

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java [hunk 8]
```diff
@@ -391,6 +401,11 @@ public void delete(MessageWriteSet messageSetToDelete) throws StoreException {
         if (value == null) {
           throw new StoreException("Cannot delete id " + info.getStoreKey() + " since it is not present in the index.",
               StoreErrorCodes.ID_Not_Found);
+        } else if (config.storeDeleteAuthorizationCheck && !Utils.validateAuthorization(value.getAccountId(),
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: same here

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java [hunk 9]
```diff
@@ -557,6 +572,10 @@ void compact(CompactionDetails details) throws IOException, StoreException {
     checkCapacityAndUpdateWriteStatusDelegate(log.getCapacityInBytes(), index.getLogUsedCapacity());
   }
 
+  public StoreConfig getStoreConfig() {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: not required anymore

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 10]
```diff
@@ -1176,16 +1206,19 @@ private void verifyPutFailure(MockId idToPut, StoreErrorCodes expectedErrorCode)
...
+        new MessageInfo(idToDelete, DELETE_RECORD_SIZE, idToDelete.getAccountId(), idToDelete.getContainerId(),
+            System.currentTimeMillis());
     MessageWriteSet writeSet =
         new MockMessageWriteSet(Collections.singletonList(info), Collections.singletonList(ByteBuffer.allocate(1)));
-    try {
-      store.delete(writeSet);
-      fail("Store DELETE should have failed");
-    } catch (StoreException e) {
-      assertEquals("Unexpected StoreErrorCode", expectedErrorCode, e.getErrorCode());
+    if (store.getStoreConfig().storeDeleteAuthorizationCheck == true) {
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: why is this condition required?

File: ambry-store/src/test/java/com.github.ambry.store/MockId.java [hunk 11]
```diff
@@ -52,7 +61,17 @@ public String getLongForm() {
 
   @Override
   public short sizeInBytes() {
-    return (short) (Id_Size_In_Bytes + id.length());
+    return (short) (Id_Size_In_Bytes + id.length() + Short.BYTES + Short.BYTES);
+  }
+
+  @Override
+  public short getAccountId() {
+    return accountId;
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/MockId.java]: I think its safer to return -1 here instead of generating random account and container IDs

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 12]
```diff
@@ -868,6 +868,14 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
     }
   }
 
+  public static boolean validateAuthorization(short storeAccountId, short storeContainerId, short requestAccountId,
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: You can move this function to a `private` function in `BlobStore`

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 10 above)
Comment (zzmao) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: The result of below check depends on the value of storeDeleteAuthorizationCheck in StoreCofing.java.

Since we may change the config value, I added the if condition.

File: ambry-store/src/test/java/com.github.ambry.store/MockId.java (hunk 11 above)
Comment (zzmao) [reply ambry-store/src/test/java/com.github.ambry.store/MockId.java]: Storing random account and container ID makes the test strict. 
If code changes impact account/container validation, it helpful to find out if the change is good.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java (hunk 7 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: Sounds good. I will use "metrics.getAuthorizationFailure.inc();", which is also used by AmbryRequest.

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 13]
```diff
@@ -570,13 +572,84 @@ public void putErrorCasesTest() {
...
-  public void deleteErrorCasesTest() {
+  public void deleteErrorCasesTest() throws Exception {
     // ID that is already deleted
     verifyDeleteFailure(deletedKeys.iterator().next(), StoreErrorCodes.ID_Deleted);
     // ID that does not exist
     verifyDeleteFailure(getUniqueId(), StoreErrorCodes.ID_Not_Found);
   }
 
+  /**
+   * Test return value for {@link BlobStore#validateAuthorization(short, short, short, short)} )}.
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: We will need to rewrite this test once we make that function `private boolean`

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 14]
```diff
@@ -868,6 +868,8 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
     }
   }
 
+
```
Comment (vgkholla) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: minor: can you make sure there are no changes at all in this file?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 15]
```diff
@@ -570,13 +572,84 @@ public void putErrorCasesTest() {
...
+   */
+  @Test
+  public void getAuthorizationFailureTest() throws Exception {
+    MockId addedId = put(1, PUT_RECORD_SIZE, Utils.Infinite_Time).get(0);
+    assertNotSame("accountId should not be -1 for added Blob.", -1, addedId.getAccountId());
+    assertNotSame("containerId should not be -1 for added Blob.", -1, addedId.getContainerId());
+    if (store.getStoreConfig().storeValidateAuthorization == true) {
+      verifyGetFailure(new MockId(addedId.getID()), StoreErrorCodes.Authorization_Failure);
+    } else {
+      store.get(Collections.singletonList(addedId), EnumSet.noneOf(StoreGetOptions.class));
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: why do this at all? To make sure it succeeds? Shouldn't we check that the return is ok in that case?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java (hunk 9 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: this needs to be removed

File: ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java [hunk 16]
```diff
@@ -136,6 +138,10 @@ public StoreMetrics(String prefix, MetricRegistry registry) {
         registry.histogram(MetricRegistry.name(IndexSegment.class, name + "SegmentsAccessedPerBlobCount"));
     identicalPutAttemptCount =
         registry.counter(MetricRegistry.name(PersistentIndex.class, name + "IdenticalPutAttemptCount"));
+    getAuthorizationFailureCount =
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java]: the owner class of the metrics should be `BlobStore.class`

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 17]
```diff
@@ -577,6 +577,86 @@ public void deleteErrorCasesTest() {
...
+  public void deleteAuthorizationSuccessTest() throws Exception {
+    short[] accountIds = {-1, Utils.getRandomShort(TestUtils.RANDOM), -1};
+    short[] containerIds = {-1, -1, Utils.getRandomShort(TestUtils.RANDOM)};
+    for (int i = 0; i < accountIds.length; i++) {
+      MockId mockId = putSingleBlob(UtilsTest.getRandomString(10), accountIds[i], containerIds[i]);
+      MessageInfo info = new MessageInfo(mockId, DELETE_RECORD_SIZE, Utils.getRandomShort(TestUtils.RANDOM),
+          Utils.getRandomShort(TestUtils.RANDOM), System.currentTimeMillis());
+      MessageWriteSet writeSet = new MockMessageWriteSet(Collections.singletonList(info),
+          Collections.singletonList(ByteBuffer.allocate(DELETE_RECORD_SIZE)));
+      store.delete(writeSet);
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: you aren't changing the account and container ID in the ID though, should you be doing that to make this test relevant (like you do in the equivalent test in GET)?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 18]
```diff
@@ -577,6 +577,86 @@ public void deleteErrorCasesTest() {
...
+      verifyDeleteFailure(mockId, StoreErrorCodes.ID_Deleted);
+    }
+  }
+
+  /**
+   * Test DELETE authorization failure for {@link BlobStore#delete(MessageWriteSet)}.
+   */
+  @Test
+  public void deleteAuthorizationFailureTest() throws Exception {
+    if (store.getStoreConfig().storeValidateAuthorization == true) {
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: you can use `assumeTrue(store.getStoreConfig().storeValidateAuthorization)` - the test won't run if it is not true (will get skipped, won't fail).

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 19]
```diff
@@ -805,6 +885,17 @@ private MockId getUniqueId() {
     return ids;
   }
 
+  private MockId putSingleBlob(String key, short accountId, short containerId) throws StoreException {
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: is it possible to modify the `put()` function in this class instead of writing a new one? If you use that function, you can even check the `storeInfo` that gets returned on GET

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 20]
```diff
@@ -577,6 +577,86 @@ public void deleteErrorCasesTest() {
...
+   * If either accountId or containerId is -1, validation is skipped.
+   */
+  @Test
+  public void getAuthorizationSuccessTest() throws Exception {
+    short[] accountIds = {-1, Utils.getRandomShort(TestUtils.RANDOM), -1};
+    short[] containerIds = {-1, -1, Utils.getRandomShort(TestUtils.RANDOM)};
+    for (int i = 0; i < accountIds.length; i++) {
+      MockId mockId = putSingleBlob(UtilsTest.getRandomString(10), accountIds[i], containerIds[i]);
+      store.get(Collections.singletonList(mockId), EnumSet.noneOf(StoreGetOptions.class));
+      store.get(
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: see my comment about modifying the `put()` function in this class. If you do that, you can use `checkStoreInfo` to make sure the return from these gets is correct

File: ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java [hunk 21]
```diff
@@ -35,6 +35,7 @@
     tempMap.put(StoreErrorCodes.ID_Not_Found, ServerErrorCode.Blob_Not_Found);
     tempMap.put(StoreErrorCodes.TTL_Expired, ServerErrorCode.Blob_Expired);
     tempMap.put(StoreErrorCodes.Already_Exist, ServerErrorCode.Blob_Already_Exists);
+    tempMap.put(StoreErrorCodes.Authorization_Failure, ServerErrorCode.Blob_Not_Found);
```
Comment (cgtz) [inline ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java]: I suppose this is okay for now, but should we eventually add a new `ServerErrorCode` for auth failure and add handling for it in the router? Right now, this could make the router send a request to every replica.

Comment (cgtz) [reply ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java]: There also seems to be a `ServerErrorCode.Bad_Request` code. @vgkholla do you think we could use that one here?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 19 above)
Comment (zzmao) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: The problem here is we want specify accoundId and containerId. We need to add two arguments for put().

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java [hunk 22]
```diff
@@ -14,6 +14,8 @@
 package com.github.ambry.store;
 
 import com.codahale.metrics.Timer;
+import com.github.ambry.account.Account;
```
Comment (cgtz) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: Are these 2 imports needed?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 18 above)
Comment (zzmao) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: nice.

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 23]
```diff
@@ -577,6 +577,86 @@ public void deleteErrorCasesTest() {
...
+          EnumSet.noneOf(StoreGetOptions.class));
+    }
+  }
+
+  /**
+   * Test GET authorization failure for {@link BlobStore#get(List, EnumSet)}.
+   */
+  @Test
+  public void getAuthorizationFailureTest() throws Exception {
+    if (store.getStoreConfig().storeValidateAuthorization == true) {
```
Comment (cgtz) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: You can use assumeTrue here too. See Gopal's comment above.

File: ambry-api/src/main/java/com.github.ambry/store/StoreKey.java [hunk 24]
```diff
@@ -37,6 +37,18 @@
    */
   public abstract String getID();
 
+  /**
+   * Get accountId of StoreKey, return -1 if unknown.
+   * @return accountId
+   */
+  public abstract short getAccountId();
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKey.java]: Could you add `@overrides` to the corresponding methods in the `BlobId` class?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 17 above)
Comment (zzmao) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: You mean try to do store.delete() with different combination of accountId/containerId, right?

File: ambry-api/src/main/java/com.github.ambry/store/StoreKey.java (hunk 24 above)
Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/store/StoreKey.java]: yes.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java (hunk 22 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: yes. In function validateAuthorization()

File: ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java (hunk 21 above)
Comment (vgkholla) [reply ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java]: @cgtz that's a good point - we certainly don't want to waste resources on that. I have a feeling that `BadRequest` might turn into an `InternalServerError` at the `Router` since the `Router` is not supposed to be making any bad requests. The other option is to make a new error code for this and then get the router to understand it. However, that means that the new router code should be deployed before this is (which might be ok). Perhaps we can do that in a follow up PR (immediately after). @zzmao what do you think?

Comment (vgkholla) [reply ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java]: To be clear, I suggest we keep this code right now and change the mapping in the follow up PR

File: ambry-store/src/main/java/com.github.ambry.store/BlobStore.java (hunk 9 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStore.java]: removed.

File: ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java (hunk 21 above)
Comment (zzmao) [reply ambry-server/src/main/java/com.github.ambry.server/ErrorMapping.java]: Using Blob_Not_Found in this PR. Issue #860 is created for the following changes.
